source/client.c

```diff
@@ -173,6 +173,54 @@ static void s_init_statistics(struct aws_mqtt_connection_operation_statistics_im
     aws_atomic_store_int(&stats->unacked_operation_size_atomic, 0);
 }
 
+static bool s_is_topic_shared_topic(struct aws_byte_cursor *input) {
+    char *input_str = (char *)input->ptr;
+    if (strncmp("$share", input_str, strlen("$share")) == 0) {
+        return true;
+    }
+    return false;
+}
+
+static struct aws_string *s_get_normal_topic_from_shared_topic(struct aws_string *input) {
+    struct aws_byte_cursor input_cursor = aws_byte_cursor_from_string(input);
+    struct aws_byte_cursor split_part;
+    AWS_ZERO_STRUCT(split_part);
+
+    // The format of a valid shared subscription is ALWAYS going to be: "$share/<group identifier>/<topic>"
+    // and therefore we must skip the first two '/' results
+    // // First call always return the whole string...
+    if (aws_byte_cursor_next_split(&input_cursor, '/', &split_part) == false) {
+        AWS_LOGF_ERROR(
+            AWS_LS_MQTT_CLIENT,
+            "Cannot parse shared subscription topic: There are no '/' in shared subscription topic filter");
+        return NULL;
+    }
+    // gets rid of "$share"
+    if (aws_byte_cursor_next_split(&input_cursor, '/', &split_part) == false) {
+        AWS_LOGF_ERROR(
+            AWS_LS_MQTT_CLIENT,
+            "Cannot parse shared subscription topic: There is no '$share/' in shared subscription topic filter");
+        return NULL;
+    }
+    // gets rid of "<group identifier>"
+    if (aws_byte_cursor_next_split(&input_cursor, '/', &split_part) == false) {
+        AWS_LOGF_ERROR(
+            AWS_LS_MQTT_CLIENT,
+            "Cannot parse shared subscription topic: There is no group identifier in shared subscription topic filter");
+        return NULL;
+    }
+    // Make sure there is something in <topic>, properly detecting "$share/<group identifier>/" as invalid
+    if (split_part.len <= 0) {
+        AWS_LOGF_ERROR(
+            AWS_LS_MQTT_CLIENT,
+            "Cannot parse shared subscription topic: There is no topic in shared subscription topic filter");
+        return NULL;
+    }
+
+    struct aws_string *result_string = aws_string_new_from_c_str(input->allocator, (char *)split_part.ptr);
```

Comment:
Check to make sure if a topic has an `/` character in it, that we aren't only returning a partial topic. I think aws_byte_cursor_next_split isn't assigning `split_part` as the entire remainder of the `input_cursor` but just the next portion of it between the split characters.